packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1228,40 +1717,336 @@ export class KeyringController extends BaseController<
+   * - Makes the first account the selected account.
+   * @param type - Keyring type to instantiate.
+   * @param opts - Optional parameters required to instantiate the keyring.
+   * @returns A promise that resolves if the operation was successful.
+   */
+  async #createKeyring(type: string, opts?: unknown) {
+    const keyring = (await this.addNewKeyring(type, opts)) as
+      | EthKeyring<Json>
+      | undefined;
+    if (!keyring) {
+      throw new Error(KeyringControllerError.NoKeyring);
+    }
+
+    const [firstAccount] = await keyring.getAccounts();
+    if (!firstAccount) {
+      throw new Error(KeyringControllerError.NoFirstAccount);
+    }
+  }
+
+  /**
+   * Instantiate, initialize and return a new keyring
+   *
+   * The keyring instantiated is of the given `type`.
+   *
+   * @param type - The type of keyring to add.
+   * @param data - The data to restore a previously serialized keyring.
+   * @returns The new keyring.
+   */
+  async #newKeyring(type: string, data: unknown): Promise<EthKeyring<Json>> {
+    const keyringBuilder = this.#getKeyringBuilderForType(type);
+
+    if (!keyringBuilder) {
+      throw new Error(
+        `${KeyringControllerError.NoKeyringBuilder}. Keyring type: ${type}`,
+      );
+    }
+
+    const keyring = keyringBuilder();
+
+    // @ts-expect-error Enforce data type after updating clients
+    await keyring.deserialize(data);
+
+    if (keyring.init) {
+      await keyring.init();
+    }
+
+    return keyring;
+  }
+
+  /**
+   * Clear Keyrings
+   *
+   * Deallocates all currently managed keyrings and accounts.
+   * Used before initializing a new vault and after locking
+   * MetaMask.
+   */
+  async #clearKeyrings() {
+    // clear keyrings from memory
+    for (const keyring of this.#keyrings) {
+      await this.#destroyKeyring(keyring);
+    }
+    this.#keyrings = [];
+    this.update((state) => {
+      state.keyrings = [];
+    });
+  }
+
+  /**
+   * Restore Keyring Helper
+   *
+   * Attempts to initialize a new keyring from the provided serialized payload.
+   * On success, returns the resulting keyring instance.
+   *
+   * @param serialized - The serialized keyring.
+   * @param serialized.type - Keyring type.
+   * @param serialized.data - Keyring data.
+   * @returns The deserialized keyring or undefined if the keyring type is unsupported.
+   */
+  async #restoreKeyring(
+    serialized: SerializedKeyring,
+  ): Promise<EthKeyring<Json> | undefined> {
+    const { type, data } = serialized;
+
+    let keyring: EthKeyring<Json> | undefined;
+    try {
+      keyring = await this.#newKeyring(type, data);
+    } catch (error) {
+      // Ignore error.
+      console.error(error);
+    }
+
+    if (!keyring) {
+      this.#unsupportedKeyrings.push(serialized);
+      return undefined;
+    }
+
+    // getAccounts also validates the accounts for some keyrings
+    await keyring.getAccounts();
+    this.#keyrings.push(keyring);
+    return keyring;
   }
```

Comment:
This is probably more readable this way:

```suggestion
  /**
   * Restore a Keyring from a provided serialized payload.
   * On success, returns the resulting keyring instance.
   *
   * @param serialized - The serialized keyring.
   * @returns The deserialized keyring or undefined if the keyring type is unsupported.
   */
  async #restoreKeyring(
    serialized: SerializedKeyring,
  ): Promise<EthKeyring<Json> | undefined> {
    try {
      const { type, data } = serialized;
      const keyring = await this.#newKeyring(type, data);

      // getAccounts also validates the accounts for some keyrings
      await keyring.getAccounts();
      this.#keyrings.push(keyring);

      return keyring;
    } catch (_) {
      this.#unsupportedKeyrings.push(serialized);
      return undefined;
    }
  }
```

Reply:
Refactored in [6153671](https://github.com/MetaMask/core/pull/3830/commits/6153671eb917062adc202d1839dfb434002f24e2) along with tests